File: term5/dc/lab1_7/go/main.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"sync"
	"time"
)

type IntVector []int
type IntMatrix []IntVector

func (matrix IntMatrix) getRows() int {
	return len(matrix)
}

func (matrix IntMatrix) getCols() int {
	return len(matrix[0])
}

func (matrix IntMatrix) isSquare() bool {
	return matrix.getRows() == matrix.getCols()
}

func newMatrix(rows, cols int) IntMatrix {
	matrix := make(IntMatrix, rows)

	for i := range matrix {
		matrix[i] = make(IntVector, cols)
	}

	return matrix
}

func generateRandomMatrix(rows, cols, min, max int) IntMatrix {
	matrix := newMatrix(rows, cols)
	random := rand.New(rand.NewSource(time.Now().UnixNano()))

	for i := 0; i < matrix.getRows(); i++ {
		for j := 0; j < matrix.getCols(); j++ {
			matrix[i][j] = random.Intn(max - min) + min
		}
	}

	return matrix
}

func logMatrix(matrix IntMatrix) {
	for i := 0; i < matrix.getRows(); i++ {
		for j := 0; j < matrix.getCols(); j++ {
			fmt.Printf("%v\t", matrix[i][j])
		}

		fmt.Println()
	}
}

func tapeMethod(lhs, rhs IntMatrix, workerCount int) IntMatrix {
	if !lhs.isSquare() || !rhs.isSquare() || lhs.getRows() != rhs.getRows() {
		panic("Bad size!")
	}

	size := lhs.getRows()

	result := newMatrix(lhs.getRows(), rhs.getCols())

	group := sync.WaitGroup{}
	group.Add(workerCount)

	for worker := 0; worker < workerCount; worker++ {
		go func(worker int) {
			pivot := int(math.Ceil(float64(size) / float64(workerCount)))
			index := worker

			for row := index * pivot; row < (index + 1) * pivot	&& row < size; row++ {
				for i := 0; i < size; i++ {
					for j := 0; j < size; j++ {
						result[row][index] += lhs[row][j] * rhs[j][index]
					}

					index = (index + 1) % size
				}
			}

			group.Done()
		}(worker)
	}

	group.Wait()

	return result
}

func getElapsedTime(function func()) float64 {
	start := time.Now()

	function()

	return float64(time.Since(start)) * 1e-6
}

func runMultiplication(matrixSize, runCount, workerCount int) float64 {
 	var sum float64 = 0

	for i := 0; i < runCount; i++ {
		lhs := generateRandomMatrix(matrixSize, matrixSize, -100, 100)
		rhs := generateRandomMatrix(matrixSize, matrixSize, -100, 100)

		elapsedTime := getElapsedTime(func() {
			tapeMethod(lhs, rhs, workerCount)
		})

		sum += elapsedTime
	}

	return sum / float64(runCount)
}

func main() {
	const runCount = 50

	logResult := func(size int) {
		worker1Result := runMultiplication(size, runCount, 1)
		worker2Result := runMultiplication(size, runCount, 2)
		worker4Result := runMultiplication(size, runCount, 4)

		fmt.Printf(
			"%3dx%-3d:\t%.10f\t%.10f\t%.10f\n",
			size,
			size,
			worker1Result,
			worker2Result,
			worker4Result,
		)
	}

	logResult(8)
	logResult(24)
	logResult(80)
	logResult(160)
	logResult(240)
	logResult(320)
	logResult(400)
}

File: term5/dc/lab1_7/java/src/main/java/Main.java
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ThreadLocalRandom;
import java.lang.Math;

public class Main {

    public static int[][] generateMatrix(final int size, final int min, final int max) {
        var matrix = new int[size][size];

        for (int i = 0; i < size; ++i) {
            for (int j = 0; j < size; ++j) {
                matrix[i][j] = ThreadLocalRandom.current().nextInt(max - min) + min;
            }
        }

        return matrix;
    }

    public static void logMatrix(final int[][] matrix) {
        for (var row : matrix) {
            for (var coefficient : row) {
                System.out.print(coefficient + "\t");
            }

            System.out.println();
        }
    }

    public static int[][] multiplyByTapeMethod(final int[][] lhs, final int[][] rhs, final int workerCount) {
        assert lhs.length == rhs.length;

        final int size = lhs.length;
        final var pool = new ForkJoinPool(workerCount);

        var result = new int[size][size];

        for (int worker = 0; worker < workerCount; ++worker) {
            int finalWorker = worker;

            pool.invoke(new RecursiveAction() {

                @Override
                protected void compute() {
                    final int pivot = (int)Math.ceil((float)size / (float)workerCount);

                    int index = finalWorker;

                    for (int row = index * pivot; row < (index + 1) * pivot	&& row < size; ++row) {
                        for (int i = 0; i < size; ++i) {
                            for (int j = 0; j < size; ++j) {
                                result[row][index] += lhs[row][j] * rhs[j][index];
                            }

                            index = (index + 1) % size;
                        }
                    }
                }
            });
        }

        return result;
    }

    public static double run(final int size, final int runCount, final int workerCount) {
        double time = 0.0;

        for (int i = 0; i < runCount; ++i) {
            final var lhs = generateMatrix(size, -100, 100);
            final var rhs = generateMatrix(size, -100, 100);

            final var begin = System.currentTimeMillis();
            {
                multiplyByTapeMethod(lhs, rhs, workerCount);
            }
            final var end = System.currentTimeMillis();

            time += end - begin;
        }

        return time / (double)runCount;
    }

    public static void logResult(final int size, final int runCount) {
        var worker1Time = run(size, runCount, 1);
        var worker2Time = run(size, runCount, 2);
        var worker4Time = run(size, runCount, 4);

        System.out.printf(
            "%3dx%-3d:\t%.10f\t%.10f\t%.10f\n",
            size,
            size,
            worker1Time,
            worker2Time,
            worker4Time
        );
    }

    public static void main(String[] args) {
        final int runCount = 50;

        logResult(8, runCount);
        logResult(24, runCount);
        logResult(80, runCount);
        logResult(160, runCount);
        logResult(240, runCount);
        logResult(160, runCount);
        logResult(320, runCount);
        logResult(400, runCount);
    }
}
